### DISCUSSION TITLE
error:020000B3:rsa routines::missing private key

### DISCUSSION THREAD

**Discussion by Dharmendra143c** [openssl/openssl#26065] [Q&A]
Hi @mattcaswell,

I was using openssl 3.0.15. With the below code the decryption is working fine. I am using public key to decrypt. it works perfectly. I am also doing the upgrade to 3.4.0 removed the deprecated api and use the latest api to do decryption but getting **error:020000B3:rsa routines::missing private key. What i want that i should be able to decrypt the data was was encrypted earlier using public key  with latest openssl 3.4**

**Working code with public decryption**
``` C++
            std::string decryptedText = "";

            RSA* rsa = NULL;
            BIO* keyBio = BIO_new_mem_buf(publicKey, 788);
            rsa = PEM_read_bio_RSA_PUBKEY(keyBio, nullptr, nullptr, nullptr);

            if (!rsa)
            {
                EXT_CONFIG_LOG_ERROR(WIDE_STRING("Error getting public key."));
                return decryptedText;
            }

            BIO_free(keyBio);

            int keySize = RSA_size(rsa);
            // Buffers for encrypted and decrypted data
            unsigned char* encryptedBuffer = new unsigned char[keySize];
            unsigned char* decryptedBuffer = new unsigned char[keySize];

            std::ifstream encrypted(encryptedFile, std::ios::binary);
            try
            {
                // Decrypt the file in chunks
                while (encrypted.read(reinterpret_cast<char*>(encryptedBuffer), keySize) &&
                    !encrypted.eof() && !encrypted.bad())
                {
                    int decryptedSize = RSA_public_decrypt(
                        keySize, encryptedBuffer, decryptedBuffer, rsa, RSA_PKCS1_PADDING);
                    if (decryptedSize == -1)
                    {
                        
                        delete[] encryptedBuffer;
                        delete[] decryptedBuffer;
                        RSA_free(rsa);
                        encrypted.close();
                        EXT_CONFIG_LOG_ERROR(WIDE_STRING("Encrypted File Closed"));
                        return decryptedText;
                    }
                    auto decryptedData =
                        std::string(reinterpret_cast<char*>(decryptedBuffer), decryptedSize);
                    decryptedText += decryptedData;
                }

                delete[] encryptedBuffer;
                delete[] decryptedBuffer;
                RSA_free(rsa);
                encrypted.close();
            }
```

**Code giving error**
``` C++
            std::string decryptedText = "";

            // Load the public key
            BIO* keyBio = BIO_new_mem_buf(publicKey, -1); // Assuming publicKey is a char array containing the public key
            if (!keyBio) {
                
                return decryptedText;
            }

            EVP_PKEY* pkey = PEM_read_bio_PUBKEY(keyBio, nullptr, nullptr, nullptr);
            BIO_free(keyBio);

            if (!pkey) {
     
                return decryptedText;
            }

            // Determine the size of the key
            int keySize = EVP_PKEY_size(pkey);

            // Buffers for encrypted and decrypted data
            unsigned char* encryptedBuffer = new unsigned char[keySize];
            unsigned char* decryptedBuffer = new unsigned char[keySize];

            std::ifstream encrypted(encryptedFile, std::ios::binary);
            if (!encrypted.is_open()) {
               
                delete[] encryptedBuffer;
                delete[] decryptedBuffer;
                EVP_PKEY_free(pkey);
                return decryptedText;
            }

            try
            {
                // Decrypt the file in chunks
                while (encrypted.read(reinterpret_cast<char*>(encryptedBuffer), keySize) &&
                    !encrypted.eof() && !encrypted.bad())
                {
                    size_t decryptedSize = keySize;
                    EVP_PKEY_CTX* ctx = EVP_PKEY_CTX_new(pkey, nullptr);
                    if (!ctx) {
                       
                        delete[] encryptedBuffer;
                        delete[] decryptedBuffer;
                        EVP_PKEY_free(pkey);
                        encrypted.close();
                        return decryptedText;
                    }

                    if (EVP_PKEY_decrypt_init(ctx) <= 0) {
                        
                        EVP_PKEY_CTX_free(ctx);
                        delete[] encryptedBuffer;
                        delete[] decryptedBuffer;
                        EVP_PKEY_free(pkey);
                        encrypted.close();
                        return decryptedText;
                    }

                    if (EVP_PKEY_CTX_set_rsa_padding(ctx, RSA_PKCS1_PADDING) <= 0) {
                       
                        EVP_PKEY_CTX_free(ctx);
                        delete[] encryptedBuffer;
                        delete[] decryptedBuffer;
                        EVP_PKEY_free(pkey);
                        encrypted.close();
                        return decryptedText;
                    }

                    if (EVP_PKEY_decrypt(ctx, decryptedBuffer, &decryptedSize, encryptedBuffer, keySize) <= 0) {
                       
                        EVP_PKEY_CTX_free(ctx);
                        delete[] encryptedBuffer;
                        delete[] decryptedBuffer;
                        EVP_PKEY_free(pkey);
                        encrypted.close();
                        return decryptedText;
                    }

                    EVP_PKEY_CTX_free(ctx);

                    auto decryptedData = std::string(reinterpret_cast<char*>(decryptedBuffer), decryptedSize);
                    decryptedText += decryptedData;
                }

                delete[] encryptedBuffer;
                delete[] decryptedBuffer;
                EVP_PKEY_free(pkey);
                encrypted.close();
```

Can you please help me identify what could be the issue.

**Comment by mattcaswell** [comment]
`EVP_PKEY_decrypt` is somewhat equivalent to the low level function `RSA_private_decrypt` and requires access to the RSA private key.

It's difficult to say for sure without understanding the full context of what you are attempting to achieve. But from the snippet of code you have shown this looks totally insecure.

The typical use case for the `RSA_public_decrypt` function that you are using is in verification of signatures, i.e. a "signature" is created by "encrypting" the data to be signed (typically a hash of a message), using `RSA_private_encrypt`, and then `RSA_public_decrypt` can be used to verify the signature by decrypting the data and confirming that the decrypted data is the expected hash of the message.

You seem to be attempting to use these functions to "decrypt" a file. That's a very strange thing to do using `RSA_public_decrypt`. Since that function works using a public key, anyone in possession of that key (which is intended to be visible to anyone) will be able to decrypt the file. It offers no confidentiality guarantees at all.

If what you really want to do is verify a signature on a whole file then you don't do it like this. You should hash the whole contents of the file, and then verify the signature on the hash. You should look at the `EVP_DigestSign*` functions (for signing) and the equivalent `EVP_DigestVerify*` functions (for verifying).

On the other hand, if what you're trying to achieve is to really encrypt/decrypt a file based on an RSA key for confidentiality purposes then you should instead consider using the CMS_* APIs for this.

Finally if you are confident that you *really* know what you are doing, and you intended to do an RSA_public_decrypt (i.e. a signature verification) then you can use `EVP_PKEY_verify` or `EVP_PKEY_verify_recover`.

**Reply by Dharmendra143c** [reply]
I have file which was encrypted using private key.
I want to decrypt the file content.  I am able to to do the the decryption using the below code.  The below code was written using openssl 3.0.15. Now i am upgrading to openssl 3.4.0. While upgrading, I am getting PEM_read_bio_RSA_PUBKEY, RSA_size, etc is deprecated.  I don't want to use the deprecated function. I want to use the new function from version 3.4.0 .This should be able to decrypt the file encrypted using older version and new version as well. The logic should be backward compatible.

``` C++
        RSA* rsa = NULL;
        BIO* keyBio = BIO_new_mem_buf(publicKey, 788);
        rsa = PEM_read_bio_RSA_PUBKEY(keyBio, nullptr, nullptr, nullptr);

        if (!rsa)
        {
            EXT_CONFIG_LOG_ERROR(WIDE_STRING("Error getting public key."));
            return decryptedText;
        }

        BIO_free(keyBio);

        int keySize = RSA_size(rsa);
        // Buffers for encrypted and decrypted data
        unsigned char* encryptedBuffer = new unsigned char[keySize];
        unsigned char* decryptedBuffer = new unsigned char[keySize];

        std::ifstream encrypted(encryptedFile, std::ios::binary);
        try
        {
            // Decrypt the file in chunks
            while (encrypted.read(reinterpret_cast<char*>(encryptedBuffer), keySize) &&
                !encrypted.eof() && !encrypted.bad())
            {
                int decryptedSize = RSA_public_decrypt(
                    keySize, encryptedBuffer, decryptedBuffer, rsa, RSA_PKCS1_PADDING);
                if (decryptedSize == -1)
                {
                    
                    delete[] encryptedBuffer;
                    delete[] decryptedBuffer;
                    RSA_free(rsa);
                    encrypted.close();
                    EXT_CONFIG_LOG_ERROR(WIDE_STRING("Encrypted File Closed"));
                    return decryptedText;
                }
                auto decryptedData =
                    std::string(reinterpret_cast<char*>(decryptedBuffer), decryptedSize);
                decryptedText += decryptedData;
            }

            delete[] encryptedBuffer;
            delete[] decryptedBuffer;
            RSA_free(rsa);
            encrypted.close();
        }
```

**Reply by mattcaswell** [reply]
> I have file which was encrypted using private key.

You do not have what you think you have. Encrypting a file with a private key makes no sense. As far as I can tell from your code, what you have done is signed the file (actually to be more accurate you have signed lots of chunks of the file, and then concatenated the signatures together). RSA implements signing via "encrypting with a private key". This isn't real encryption. Anyone with the public key (which by its nature is intended to be public) can decrypt the contents. Therefore you achieve no confidentiality from doing this.

If you want to do a verify (which in RSA is implemented via "decrypting with a public key") then, as I said in previous answer, you can use `EVP_PKEY_verify` or `EVP_PKEY_verify_recover`. But - really you need to forget about backwards compatibility and start again. If your objective is encryption/decryption then you haven't achieved it with your original code, or your new code. Throw this code away and start again (consider using the CMS_* APIs as per my earlier suggestion).
